src/execution/operator/csv_scanner/sniffer/type_detection.cpp

```diff
@@ -0,0 +1,407 @@
+					string separator;
+					bool has_format_is_set = false;
+					auto format_iterator = candidate->dialect_options.has_format.find(sql_type.id());
+					if (format_iterator != candidate->dialect_options.has_format.end()) {
+						has_format_is_set = format_iterator->second;
+					}
+					if (has_format_candidates.count(sql_type.id()) &&
+					    (!has_format_is_set || format_candidates[sql_type.id()].size() > 1) && !dummy_val.IsNull() &&
+					    StartsWithNumericDate(separator, StringValue::Get(dummy_val))) {
+						// generate date format candidates the first time through
+						auto &type_format_candidates = format_candidates[sql_type.id()];
+						const auto had_format_candidates = has_format_candidates[sql_type.id()];
+						if (!has_format_candidates[sql_type.id()]) {
+							has_format_candidates[sql_type.id()] = true;
+							// order by preference
+							auto entry = format_template_candidates.find(sql_type.id());
+							if (entry != format_template_candidates.end()) {
+								const auto &format_template_list = entry->second;
+								for (const auto &t : format_template_list) {
+									const auto format_string = GenerateDateFormat(separator, t);
+									// don't parse ISO 8601
+									if (format_string.find("%Y-%m-%d") == string::npos) {
+										type_format_candidates.emplace_back(format_string);
+									}
+								}
+							}
+							//	initialise the first candidate
+							candidate->dialect_options.has_format[sql_type.id()] = true;
+							//	all formats are constructed to be valid
+							SetDateFormat(*candidate, type_format_candidates.back(), sql_type.id());
+						}
+						// check all formats and keep the first one that works
+						StrpTimeFormat::ParseResult result;
+						auto save_format_candidates = type_format_candidates;
+						while (!type_format_candidates.empty()) {
+							//	avoid using exceptions for flow control...
+							auto &current_format = candidate->dialect_options.date_format[sql_type.id()];
+							if (current_format.Parse(StringValue::Get(dummy_val), result)) {
+								break;
+							}
+							//	doesn't work - move to the next one
+							type_format_candidates.pop_back();
+							candidate->dialect_options.has_format[sql_type.id()] = (!type_format_candidates.empty());
+							if (!type_format_candidates.empty()) {
+								SetDateFormat(*candidate, type_format_candidates.back(), sql_type.id());
+							}
+						}
+						//	if none match, then this is not a value of type sql_type,
+						if (type_format_candidates.empty()) {
+							//	so restore the candidates that did work.
+							//	or throw them out if they were generated by this value.
+							if (had_format_candidates) {
+								type_format_candidates.swap(save_format_candidates);
+								if (!type_format_candidates.empty()) {
+									SetDateFormat(*candidate, type_format_candidates.back(), sql_type.id());
+								}
+							} else {
+								has_format_candidates[sql_type.id()] = false;
+							}
+						}
+					}
```

Comment:
Can we move this into a separate function? I did not try to follow this logic; I shy away from anything date/time-related. 😅